(PE-23981) move from at-at to quartz

jonathannewman opened the pull request:
This moves the internal implementation from the no-longer maintained
at-at library to the heavily maintained quartz library.

It reimplements the group-id concept using the underlying group concept
built into quartz, and adds interfaces for retrieving the existing job
identifiers.

It is a functionally equivalent implementation.

In addition the library has been updated to require java 8.

justinstoller commented:
Is this a Johnson thing?

jonathannewman commented:
For the next Z.

jonathannewman commented on project.clj:
```diff
@@ -1,14 +1,15 @@
                  [puppetlabs/trapperkeeper]
                  [puppetlabs/i18n]
-                 [overtone/at-at "1.2.0"]]
+                 [overtone/at-at "1.2.0"]
```
I need to remove at-at here

jonathannewman commented on project.clj:
```diff
@@ -29,5 +30,5 @@
   :plugins  [[lein-parent "0.3.1"]
              [puppetlabs/i18n "0.8.0"]]
-
+  :aot [puppetlabs.trapperkeeper.services.scheduler.job]
```
Required to support gen-class

jonathannewman commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj:
```diff
@@ -19,46 +24,63 @@
-    (-> (at-at/after n job pool)
-        :id  ; return only the ID; do not leak the at-at RecurringJob instance
-        )))
+  ([n f pool]
```
I think this arity can go away

jonathannewman commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj:
```diff
@@ -19,46 +24,63 @@
-    (-> (at-at/interspaced n job pool)
-        :id  ; return only the ID; do not leak the at-at RecurringJob instance
-        )))
+  ([n f pool]
```
This arity can go away

jonathannewman commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj:
```diff
@@ -19,46 +24,63 @@
-  (at-at/stop-and-reset-pool! pool))
+(defn get-jobs-in-group
+  [^org.quartz.Scheduler pool group-id]
+  (let [group-name (str group-id)]
```
This should assume group-id is correctly formed

jonathannewman commented:
@justinstoller do you have time to look at this?  I was hoping to get it in relatively soon.

justinstoller commented:
Oh yeah, totally have time now.

jonathannewman commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj:
```diff
@@ -19,46 +24,58 @@
+(defn build-executable-job
+  ([f job-name group-name] (build-executable-job f job-name group-name {}))
+  ([f job-name group-name options]
+   (let [job (wrap-with-error-logging f)
```
Good idea.

jonathannewman commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj:
```diff
@@ -19,46 +24,58 @@
+
+         executable-job (-> (org.quartz.JobBuilder/newJob puppetlabs.trapperkeeper.services.scheduler.job)
+                            (.withIdentity job-name group-name)
+                            (.usingJobData jdm))]
```
Good call.

jonathannewman commented on test/integration/puppetlabs/trapperkeeper/services/scheduler/scheduler_service_test.clj:
```diff
@@ -169,11 +169,16 @@
 
 (defn guaranteed-start-interval-job
   ([service interval]
-    (guaranteed-start-interval-job service interval nil))
```
It was specifying a nil-group id.  I think it was leveraging the fact that in the  old implementation nil group-ids were the same as not using the interface that allowed group-id specification.

jonathannewman commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj:
```diff
@@ -19,46 +24,58 @@
-    (-> (at-at/interspaced n job pool)
-        :id  ; return only the ID; do not leak the at-at RecurringJob instance
-        )))
+  [n f ^org.quartz.Scheduler pool group-name]
```
That sounds like a good idea.

jonathannewman commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj:
```diff
@@ -19,46 +24,58 @@
-  [id pool]
-  (at-at/stop id pool))
+  [^org.quartz.JobKey id ^org.quartz.Scheduler pool]
+  (.deleteJob pool id))
```
It ignores jobs that don't exist, and only throws when there is an internal Quartz error.  It actually returns true if the job was deleted, and false otherwise.  I'll update the docstring.

jonathannewman commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj:
```diff
@@ -19,46 +24,58 @@
+  [^org.quartz.Scheduler pool]
+  (let [groups (seq (.getJobGroupNames pool))
+        extract-keys (fn [group-name] (seq (.getJobKeys pool (GroupMatcher/jobGroupEquals group-name))))]
+    (flatten (map extract-keys groups))))
```
Good call.

jonathannewman commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_service.clj:
```diff
@@ -67,57 +38,60 @@
   (interspaced [this n f group-id]
-               (let [id (core/interspaced n f (get-pool this))]
-                 (enqueue-job! this id {:group-id group-id})))
+    (core/interspaced n f (get-pool this) (safe-group-id group-id)))
```
I was sort of two minds about it.  On one side we don't _technically_ expose the internals, as we don't describe what it is, and this is similar to what we did with at-at.  All we suggest is that what is returned is an identifier for the job.

I could put it under a map with the name / group and create the job-key.

I think we should ask ourselves what are the downsides of using the job-key directly?  I wouldn't expect any of the clients to be persisting any of them, and they are just used during the lifetime of the application.

jonathannewman commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_service.clj:
```diff
@@ -67,57 +38,60 @@
+   (try
+     (core/get-all-jobs (get-pool this))
+     (catch SchedulerException _
+       ;; happens when shutdown
```
Good call.  The scheduler does "validateState" on every interface, and every one of them will throw the exception if it is shutdown.  There is an explicit check for "isShutdown", but that doesn't solve the race.  I'll patch each one, and add a comment as to why.

jonathannewman replied:
https://github.com/quartz-scheduler/quartz/blob/9f9e400733f51f7cb658e3319fc2c140ab8af938/quartz-core/src/main/java/org/quartz/core/QuartzScheduler.java#L762-L768

justinstoller commented on src/puppetlabs/trapperkeeper/services/scheduler/job.clj:
```diff
@@ -0,0 +1,32 @@
+              scheduler (.getScheduler context)
+              oldTrigger (.getTrigger context)
+              trigger (-> (.getTriggerBuilder oldTrigger)
+                          (.startAt (org.quartz.DateBuilder/futureDate interval org.quartz.DateBuilder$IntervalUnit/MILLISECOND))
```
more imports?

justinstoller commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj:
```diff
@@ -1,12 +1,20 @@
+    (.start scheduler)
+    scheduler))
 
 (defn wrap-with-error-logging
```
This is no longer used.

jonathannewman replied:
I didn't know if anything outside the library might be using it since it is a utility thing.  I'll take it out and ask forgiveness later. :)

justinstoller commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj:
```diff
@@ -19,46 +27,85 @@
+(defn build-executable-job
+  ([f job-name group-name] (build-executable-job f job-name group-name {}))
+  ([f job-name group-name options]
+   (let [jdm (org.quartz.JobDataMap.)
```
import nit.

justinstoller commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj:
```diff
@@ -19,46 +27,85 @@
+    (try
+      (doseq [job (get-all-jobs scheduler)]
+        (.deleteJob scheduler job))
+      (.shutdown scheduler true)
```
This will wait until jobs have completed, right? But if there's an Exception raised when calling delete on any job we'll skip this?

We might want to catch SchedulerException in the doseq loop and log a message about failing to stop job <foo>. And then call shutdown in a timeout and if the timeout elapses tell it to shutdown w/o waiting? I don't know about the reliability of shutting down Quartz, but I know we've done similar to the timeout thing in file-sync client.

jonathannewman replied:
Good point.  I'll fix it.

jonathannewman commented:
@justinstoller Good call on the shutdown behavior.  It wasn't behaving quite correctly.  I had to add the notion of the "current thread" to the job instance to allow the interrupt to be called when a job is forcibly being shut down.  There is a test to demonstrate that.

This also addresses a problem where restarting might end up with the same scheduler instance (because the jobs weren't finished).  Now, the instances are all uniquely named within Quartz, and we explicitly remove the old ones from the Quartz registry to help prevent leaks.

Let me know if any of that doesn't make sense.

jonathannewman commented on src/puppetlabs/trapperkeeper/services/scheduler/job.clj:
```diff
@@ -0,0 +1,47 @@
+
+(defn -interrupt
+  [this]
+  (println "interrupt!")
```
Oops

jonathannewman replied:
Fixed.

justinstoller commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj:
```diff
@@ -1,64 +1,124 @@
+    ; is to lookup the scheduler in the SchedulerRepository and reuse the existing one first
+    (System/setProperty "org.quartz.scheduler.instanceName" (make-instance-name next-instance-count))
+    ; explicitly remove any old instance from the scheduler repository to prevent leaks
+    (.remove sr (make-instance-name (- next-instance-count 1)))
```
I'm grokking the pros and cons to doing this here vs in stop-all-jobs, which, since it shutdown the scheduler instance is where I'd figure I'd see it removed from the registry.

jonathannewman replied:
Good point.

jonathannewman commented:
@justinstoller   I moved the naming specifically to the "stop-all-jobs" function which removed the need for the extra atom, and renaming the scheduler.  That's a lot more simple.

jonathannewman commented:
I found one case that isn't handled the same way. 
```
2018-10-08 17:34:23,537 ERROR [qtp12062675-11103] p.o.api [services.clj:1] Exception while handling HTTP POST request at /orchestrator/v1/command/schedule_task
java.lang.IllegalArgumentException: Value out of range for int: 21026149688
	at clojure.lang.RT.intCast(RT.java:1241)
	at clojure.lang.RT.intCast(RT.java:1211)
	at puppetlabs.trapperkeeper.services.scheduler.scheduler_core$after.invokeStatic(scheduler_core.clj:54)
	at puppetlabs.trapperkeeper.services.scheduler.scheduler_core$after.invoke(scheduler_core.clj:49)
	at puppetlabs.trapperkeeper.services.scheduler.scheduler_service$reify__46527$service_fnk__9422__auto___positional$reify__46532.after(scheduler_service.clj:61)
	at puppetlabs.orchestrator.scheduled.CurrentJobPool.add_job(scheduled.clj:144)
	at puppetlabs.orchestrator.scheduled$schedule_task$fn__135633$fn__135635.invoke(scheduled.clj:186)
	at puppetlabs.orchestrator.scheduled$schedule_task$fn__135633.invoke(scheduled.clj:185)
```
I don't think that is the end of the world.  We are going to improve the interface to allow different numeric equivalents to be passed in, so this is a temporary problem.

jonathannewman commented:
Actually, I take that back.  2,147,483,647 milliseconds is only about 24 days.  I'm going have to refactor the interface to take a long and convert down.

jonathannewman commented:
I changed the future date calculation, and it now supports milliseconds in the long format.  I added some tests that schedule things for far in the future that failed before.

jonathannewman commented:
I have tested this in both the component unit/integration tests as well as in the components in a build.  I haven't seen any issues.  I also scheduled tasks very far in the future, and the calculations are correct.

jonathannewman commented on test/integration/puppetlabs/trapperkeeper/services/scheduler/scheduler_service_test.clj:
```diff
@@ -3,15 +3,14 @@
       (testing "interspaced"
         (let [service (tk/get-service app :SchedulerService)
-              num-runs 3 ; let it run a few times, but not too many
+              num-runs 10 ; let it run a few times, but not too many
```
The routine is checking to make sure the interval period was honored.  Seemed like more is better in this circumstance, as 3 is just two repeat iterations.

justinstoller commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj:
```diff
@@ -1,64 +1,117 @@
+  "Returns true, if the job was deleted, and false if the job wasn't found."
+  [^JobKey id ^Scheduler scheduler]
+  (try
+    (.deleteJob scheduler id)
```
Should we interrupt this job as well?

Using stop-job in stop-all-jobs' do-seq would be nice, unless there's a good reason to keep them separate, in which case a comment would be great.

jonathannewman replied:
The documented behavior of the service's `stop-job` is that it doesn't interrupt a job that is running.  I could augment the function with a flag to indicate if it should interrupt or not and reuse the function that way.  Thoughts?

justinstoller replied:
Oh yeah, that makes sense.

I don't think I'll remember that in another year when we come back to this next, but now that you say it I see how that was the meaning of the docs.

I don't think it's necessarily worth changing the arity for different consumers.

jonathannewman replied:
We are planning on making some new additions to the interface later, and can look at exposing a flag to force interrupt at that point.  That would allow us to reuse the function more easily.
